blockchain/chain.go

```diff
@@ -1067,6 +1246,10 @@ func (b *BlockChain) reorganizeChainInternal(target *blockNode) error {
 			return err
 		}
 
+		// Ensure the associated spend journal of the disconnected block is not
+		// removed since it might be needed in a recovery process.
+		b.pruneFilter.AddEntry(block.Hash())
```

Comment:
I think it would be helpful to add more of a description around who should be calling `AddEntry`/`RemoveEntry`, and when.

Since `AddEntry` increments the number of references to a block hash that we don't want to remove, shouldn't it be called by (or incremented based on) whoever still needs it (the indexers in this case).  It's being called every time here, which probably doesn't make sense if indexes aren't enabled (we would never remove the spend journal then).  Also, if another consumer of the spend journal used this, we'd need to increment it twice (for each consumer) rather than just having blockchain increment it once in the general case.

Reply:
`AddEntry` is being called for `reorganizeChainInternal` because it's possible blocks initially disconnected the reorg process get reconnected. It's ideal to keep the spend journal data around for such a possibility. It is also possible to have out-of-sync disabled indexers that will need the spend journal entries of the disconnected blocks to recover up when enabled. It is also possible the reorganization was initiated by a call to `invalidateblock`, the spend journals for the disconnected blocks need to be kept around on the possibility the invalidated block is validated again via `reconsiderblock`. `reorganizeChainInternal` is the intend caller for `AddEntry` based on the reasons given above.

Regarding not immediately pruning the spend journal of indexes are not enabled, the cases not involving indexes mentioned above still apply. Also it is possible to have indexes disabled and out of sync. Once they are enabled they will still need the spend journal data of the disconnected blocks to recover back to the main chain. I will refine the check to exclude cases where the indexes do not exist yet but the other cases outlined earlier would still apply. 

I will update the comments for `AddEntry/RemoveEntry` as requested and take another look at improvements. I reckon possible events needing the spend journal data can collectively count as a single `AddEntry` call whereas confirmed out of sync indexes would be another another `AddEntry` call during a reorg.

Reply:
> AddEntry is being called for reorganizeChainInternal because it's possible blocks initially disconnected the reorg process get reconnected. It's ideal to keep the spend journal data around for such a possibility.

> It is also possible the reorganization was initiated by a call to invalidateblock, the spend journals for the disconnected blocks need to be kept around on the possibility the invalidated block is validated again via reconsiderblock. reorganizeChainInternal is the intend caller for AddEntry based on the reasons given above.

This is not accurate given that the spend journal is currently always removed when the block is disconnected.  The spend journal isn't needed to reconnect a block. 

My understanding of the problem that is trying to be solved here is:

- We are adding an asynchronous consumer of block connect/disconnect notifications (possibly multiple consumers in the future)
- The consumer needs the spend journal entry for disconnected blocks, but currently, that is removed immediately when disconnecting a block and is therefore not available

Given that, I think the goal of an ideal solution for this would be:

- The spend journal is retained **only** while there are still consumers that need it

In order to accomplish this, I think we would need to know whether each consumer has seen (or been notified) of the block being disconnected.  To build on your current solution, I think we could do something like:

- When a block is disconnected, "ask" each consumer if they still need the block
- If the consumer needs the block (i.e., it saw the block and hasn't disconnected it yet), increment the count
  - This should address the various cases outlined such as
    - Indexes never enabled (the disconnected block isn't needed)
    - Indexes enabled, then disabled (the disconnected block is needed only if the indexers connected that block previously)
- When that count goes to zero, remove the spend journal
- One caveat is this would require the consumer to track block hashes (which currently isn't an issue since the tx index tracks it and the addr index depends on the tx index)

The other thing that needs to be considered is how shutdowns are handled.  I suspect we may need to be tracking this in the database.